lib/internal/modules/esm/resolve.js

```diff
@@ -374,18 +374,21 @@ function finalizeResolution(resolved, base, preserveSymlinks) {
       resolved.pathname, 'must not include encoded "/" or "\\" characters',
       fileURLToPath(base));
 
-  const path = fileURLToPath(resolved);
+  let path = fileURLToPath(resolved);
   if (getOptionValue('--experimental-specifier-resolution') === 'node') {
     let file = resolveExtensionsWithTryExactName(resolved);
-    if (file !== undefined) return file;
-    if (!StringPrototypeEndsWith(path, '/')) {
-      file = resolveDirectoryEntry(new URL(`${resolved}/`));
-      if (file !== undefined) return file;
-    } else {
-      return resolveDirectoryEntry(resolved) || resolved;
+
+    if (file === undefined) {
+      file = StringPrototypeEndsWith(path, '/') ?
+        resolveDirectoryEntry(resolved) : resolveDirectoryEntry(new URL(`${resolved}/`));
+
+      if (file === undefined) {
+        throw new ERR_MODULE_NOT_FOUND(
+          resolved.pathname, fileURLToPath(base), 'module');
+      }
     }
-    throw new ERR_MODULE_NOT_FOUND(
-      resolved.pathname, fileURLToPath(base), 'module');
+
+    path = file;
```

Comment:
So it seems like the significant change here is to remove the `return` statements from the above code and instead set the resolved file to the `path` variable, which gets further attention related to symlinks below? Yet there’s also a check that the resolved `path` isn’t a directory, which makes sense for ESM resolution but not for the legacy CommonJS style; though by that point any “directory” imports would have been resolved to `<whatever>/index.js`, I assume?

Reply:
I checked that there is a chance to return a directory (when the path ends with `/` and `resolveDirectoryEntry` return undefined):

https://github.com/nodejs/node/blob/29bb2bb57d2a993e4e79172fe6b954f8047eee81/lib/internal/modules/esm/resolve.js#L381-L386

And it seems to be added explicitly:
https://github.com/nodejs/node/commit/f8976a76bbb49d678056d6d2ddd57064cf9e63c2#diff-b4c24f634e3741e5ad9e8c29864a48f2bd284a9d66d8fed3d077ccee0c44087bR253

I will follow this and jump the check below, only process symlinks.

Reply:
Can anyone remember the use case of return `resolved` (which is a directory that ends with `/`) directly?
https://github.com/nodejs/node/blob/29bb2bb57d2a993e4e79172fe6b954f8047eee81/lib/internal/modules/esm/resolve.js#L385

I can add a test case of that.

Reply:
i mean, a path of `./foo/` could be `./foo/index`, or the "main" in `./foo/package.json` - is that what you mean?

Reply:
This is handled by `resolveDirectoryEntry`.
What I wonder is the case: if `./foo/index` etc. not found, the origin `resolved` (which in case is `./foo/`) is returned.

But anyway, I kept this behavior unchanged in the latest commit, only handling file's symlink.

Reply:
I’m not sure I follow. My question is whether this breaks directory imports. Previously, the `node` mode never got down to the `resolveDirectoryEntry` logic because of the `return`s that this PR removes. Now the `node` mode does go through that block (that begins with `if (stats.isDirectory()) {`), so my question is, is this okay? I assume it is, because by the time it gets there, any specifiers of `'./folder'` have been resolved to `'./folder/index.js'`; but can you confirm that I’m correct here?

If so, we have nothing to worry about and this is fine to merge in; but if I’m wrong, we’ll need to address this case somehow.

Reply:
This change will not break directory imports.

`resolveExtensionsWithTryExactName` and `resolveDirectoryEntry` will return either a file or undefined, and will never return a directory.

If `resolveExtensionsWithTryExactName` return a file,
- before the change: it will return directly
- after the change: it will go through `isDerectory()` and `!isFile()` (which is ok) and `realpathSync` (which is we want)

If `resolveExtensionsWithTryExactName` return undefined,
- before the change: it will call `resolveDirectoryEntry` and try to find the file from package.json, if the path ends with '/' and a file is not found, it will return a dirctory.
- after the change: it will behave the same as before, then
  - if the path ends with '/' and a file is not found, it will return a directory directly.
  - if the path is not undefined (in this case, can be sure is a file), it will go through `isDerectory()` and `!isFile()` (which is ok) and `realpathSync` (which is we want).

Reply:
Excellent, thanks for checking!